aws_lambda_builders/workflows/java_maven/DESIGN.md

```diff
@@ -0,0 +1,94 @@
+# Java - Maven Lambda Builder
+
+## Scope
+
+This package enables the creation of a Lambda deployment package for Java
+projects managed using the Maven build tool.
+
+## Challenges
+
+- Java Version compatibility mentioned in the [Gradle Lambda Builder] Design doc.
+
+- Building Multimodule project (out of scope for the current version)
+
+Here `ProjectA` is a a single lambda function, and `ProjectB` is a multimodule
+project where sub modules `lambda1` and `lambda2` are each a lambda
+function. In addition, suppose that `ProjectB/lambda1` has a dependency on its
+sibling module `ProjectB/common`.
+
+**Single-module Project A**
+```
+ProjectA
+├── pom.xml
+├── src
+└── template.yaml
+```
+
+**Multi-module Project B**
+```
+ProjectB
+├── common
+│   └── pom.xml
+├── lambda1
+│   └── pom.xml
+├── lambda2
+│   └── pom.xml
+├── pom.xml
+└── template.yaml
+```
+
+Building Project A is relatively simple since we just need to issue `mvn
+package` and place the built classes and dependency jars within the artifact directory.
+
+Building `ProjectB/lambda1` requires maven to build `lambda1` module from
+the root pom directory and use `--also-make` option to build necessary dependencies 
+(`ProjectB/common`  in this case) first before building `ProjectB/lambda1`. This is because
+maven is not able to find its way back up to the parent `ProjectB` to
+also build `ProjectB/common`. The challenge part here is to find the parent pom directory 
+especially for the projects with multiple level of submodules. Building multi-module project is
+out of scope for the current version.
+
+## Implementation
+
+### Build Workflow
+
+#### Step 1: Copy source project to scratch directory
+
+By default, Maven stores its build-related metadata in a `target`
```

Comment:
This will cause slowdown on large projects, is there any other alternative to this?

Reply:
Unfortunately no. Maven doesn't support changing the `target` directory from command line. We'd need to update the pom.xml file to change it.

Reply:
One crazy idea: Can the workflow generate a pom.xml which customizes the build directory folder [1] and somehow "include" the Lambda function's pom.xml? The workflow can generate this file within scratch directory, and run `mvn` on this pom which would also run on the function's pom.

I know the gradle workflow does something similar. So is this possible?

[1] Generated pom.xml would have something like this https://stackoverflow.com/a/12598554

Reply:
Technically we could do this by injecting a ` <profile>` to a copy of customer's pom.xml file. However, unlike Gradle, Maven doesn't support it natively, i.e, we'd need to do some xml parsing and validating in order to insert the xml blob to the right place. Not sure if that's what we want.

Reply:
- Is generating the a pom to be the parent of function's pom not possible?
- Do we really need to parse the XML? Can we insert something to the end-ish by using regex to find the end tag? 

I really don't want to be in the business of parsing XMLs, but I want to make sure we have exhausted all possibilities before we think about it.

Reply:
> Is generating the a pom to be the parent of function's pom not possible?

There are more files to update with this approach.

- If the project is single-module, we need to add the following module to the customer's pom.
```
    <parent>
        <artifactId>fake-parent</artifactId>
        <groupId>groupId</groupId>
        <version>VERSION</version>
    </parent>
```
- If the project is multi-module project itself, on top of what we do in the above, we'd also need to update the project's original parent pom with
```
 <modules>
        <module>fake-parent</module>
<modules>
```

> Do we really need to parse the XML? Can we insert something to the end-ish by using regex to find the end tag?

It requires some parsing.  `<profile/>` has to be inside of `<profiles/>`, we need to firstly check if customer has profiles and if yes, insert it after the last `</profile>` tag and before `</profiles>`

If not, we need to insert 
```
<profiles>
  <profile>...</profile>
</profiles>
```

Reply:
We can have more discussions on this and optimize this later and I don't think it should be be blocker for the java maven builder. What do you think?

Reply:
Understood. Let's skip this optimization for now. I was just trying to see if there are other options. We can always make it better later